trin-core/src/portalnet/overlay.rs

```diff
@@ -361,70 +377,75 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
         enr: Enr,
         content_keys_offered: Vec<TContentKey>,
     ) -> anyhow::Result<Accept> {
-        let connection_id = response.connection_id.to_be();
+        let conn_id = response.connection_id.to_be();
 
         // Do not initialize uTP stream if remote node doesn't have interest in the offered content keys
         if response.content_keys.is_zero() {
             return Ok(response);
         }
 
-        self.utp_listener
-            .write()
-            .await
-            .listening
-            .insert(connection_id, UtpMessageId::OfferStream);
+        let utp_request = UtpListenerRequest::OfferStream(conn_id);
+        if let Err(err) = self.utp_listener_tx.send(utp_request) {
+            return Err(anyhow!("Unable to send uTP Offer stream request: {err}"));
+        }
 
         // initiate the connection to the acceptor
-        let mut conn = self
-            .utp_listener
-            .write()
-            .await
-            .connect(connection_id, enr.node_id())
-            .await?;
-
-        // Handle STATE packet for SYN
-        let mut buf = [0; BUF_SIZE];
-        conn.recv(&mut buf).await?;
-
-        // Return to acceptor: the content key and corresponding data
-        let mut content_items: Vec<(Vec<u8>, Vec<u8>)> = vec![];
-
-        for (i, key) in response
-            .content_keys
-            .clone()
-            .iter()
-            .zip(content_keys_offered.iter())
-        {
-            if i == true {
-                match self.storage.read().get(&key.clone()) {
-                    Ok(content) => match content {
-                        Some(content) => content_items.push((key.clone().into(), content)),
-                        None => {}
-                    },
-                    // should return some error if content not found
-                    Err(_) => {}
+        let (tx, rx) = tokio::sync::oneshot::channel::<anyhow::Result<UtpSocket>>();
+
+        let _ = self
+            .utp_listener_tx
+            .send(UtpListenerRequest::Connect(conn_id, enr.node_id(), tx));
+
+        match rx.await? {
+            Ok(mut conn) => {
+                // Handle STATE packet for SYN
+                let mut buf = [0; BUF_SIZE];
+                conn.recv(&mut buf).await?;
+
+                // Return to acceptor: the content key and corresponding data
+                let mut content_items: Vec<(Vec<u8>, Vec<u8>)> = vec![];
+
+                for (i, key) in response
+                    .content_keys
+                    .clone()
+                    .iter()
+                    .zip(content_keys_offered.iter())
+                {
+                    if i == true {
+                        match self.storage.read().get(&key.clone()) {
+                            Ok(content) => match content {
```

Comment:
Most of this logic didn't change, but got caught in the diff anyway because of an indent. Sometimes that's a clue to me that breaking down into smaller functions will make future reviews easier. So 👍🏻 👍🏻  if you feel up to doing that now.